File: hedera-mirror-rest/model/tokenBalance.js
// SPDX-License-Identifier: Apache-2.0

class TokenBalance {
  static tableAlias = 'tb';
  static tableName = 'token_balance';
  static TOKEN_ID = `token_id`;
  static ACCOUNT_ID = `account_id`;
  static BALANCE = 'balance';
  static CONSENSUS_TIMESTAMP = 'consensus_timestamp';

  /**
   * Gets full column name with table alias prepended.
   *
   * @param {string} columnName
   * @private
   */
  static getFullName(columnName) {
    return `${this.tableAlias}.${columnName}`;
  }
}

export default TokenBalance;

File: common/src/main/java/org/hiero/mirror/common/domain/balance/TokenBalance.java
// SPDX-License-Identifier: Apache-2.0

package org.hiero.mirror.common.domain.balance;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonUnwrapped;
import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import jakarta.persistence.EmbeddedId;
import jakarta.persistence.Entity;
import java.io.Serializable;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hiero.mirror.common.domain.entity.EntityId;
import org.springframework.data.domain.Persistable;

@Builder
@Data
@AllArgsConstructor
@NoArgsConstructor
@Entity
public class TokenBalance implements Persistable<TokenBalance.Id> {

    private long balance;

    @EmbeddedId
    @JsonUnwrapped
    private TokenBalance.Id id;

    @JsonIgnore
    @Override
    public boolean isNew() {
        return true; // Since we never update balances and use a natural ID, avoid Hibernate querying before insert
    }

    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    @Embeddable
    public static class Id implements Serializable {

        private static final long serialVersionUID = -8547332015249955424L;

        @Column(nullable = false, updatable = false) // set updatable = false to prevent additional hibernate query
        private long consensusTimestamp;

        @Column(nullable = false, updatable = false) // set updatable = false to prevent additional hibernate query
        private EntityId accountId;

        private EntityId tokenId;
    }
}
